Adding rancher-pod-collector script

mattmattox opened the pull request:
This script is designed to capture debug logs from the Rancher Pods. It using the following workflow.

- Tries to figure kubeconfig access (It allow the user to override the kubeconfig then tries to see if it's running inside the cluster, then tries the current loaded kubeconfig, then tries k3s/docker exec.
- Captures some data about the cluster (nodes, k8s versions, svc, etc)
- Turns on debug logging in all Rancher pods (loglevel --set debug)
- Gives the user a pause message to give them time to reproduce the issue
- Turns off debug logging in all Rancher pods (loglevel --set info)
- Capture logs for all Rancher pods
- Archive output (using the same process as log collector)

dnoland1 commented:
Script didn't work for me:
```
$ ./rancher-pod-collector.sh
./rancher-pod-collector.sh: line 43: syntax error near unexpected token `then'
./rancher-pod-collector.sh: line 43: `  then'
```

mattmattox commented:
@dnoland1 That should be fixed now.

mattmattox commented:
@superseb As requested I have added a flag `-t` to enable trace logging instead of just debug. I also added the warning message `WARNING: Trace logging has been set. Please confirm that you understand this may capture sensitive information.`

oxr463 commented on collection/rancher/v2.x/rancher-pod-collector/rancher-pod-collector.sh:
```diff
@@ -0,0 +1,217 @@
+#!/bin/bash
+SPACE="512"
+TIMEOUT="60"
```
What is this for?

mattmattox replied:
The variable `SPACE` is for the minimum free space in MBs for the tmp dir.
I removed the `TIMEOUT` that was leftover from the log script.

oxr463 commented on collection/rancher/v2.x/rancher-pod-collector/rancher-pod-collector.sh:
```diff
@@ -0,0 +1,217 @@
+      MKTEMP_BASEDIR="-p ${OPTARG}"
+      ;;
+    r)
+      RUNTIME_FLAG="${OPTARG}"
```
What is this for?

mattmattox replied:
Removing, this was leftover from the log script.

dnoland1 reviewed (changes requested):
Tested on my Mac with a kubeconfig for a v1.18.15+k3s1 cluster running Rancher v2.5.5:
```
$ ./rancher-pod-collector.sh
2021-02-02 22:21:23: Created /var/folders/3p/mtv3g5dx3cv8cgm22p1plrl00000gp/T/tmp.t8zPODjY
2021-02-02 22:21:23: Verifying cluster access
2021-02-02 22:21:25: Cluster access has been verified
2021-02-02 22:21:25: Collecting cluster info
2021-02-02 22:21:39: Enabling debug for Rancher pods
2021-02-02 22:21:41: Pod: rancher-997c9fd5b-7hljr OK
2021-02-02 22:21:42: Pod: rancher-997c9fd5b-9cnsz OK
Press any key to continue or Ctrl+C to exit...
2021-02-02 22:22:00: Disabling debug for Rancher pods
2021-02-02 22:22:02: Pod: rancher-997c9fd5b-7hljr OK
2021-02-02 22:22:03: Pod: rancher-997c9fd5b-9cnsz OK
2021-02-02 22:22:03: Capturing debug logs from Rancher pods
2021-02-02 22:22:04: Pod: rancher-997c9fd5b-7hljr
2021-02-02 22:22:05: Pod: rancher-997c9fd5b-9cnsz
2021-02-02 22:22:07: Created /var/folders/3p/mtv3g5dx3cv8cgm22p1plrl00000gp/T/david-mbp-2017.local-2021-02-02_22_22_06.tar.gz
2021-02-02 22:22:07: Removing /var/folders/3p/mtv3g5dx3cv8cgm22p1plrl00000gp/T/tmp.t8zPODjY
```
This file didn't seem to get generated correctly:
```
$ cat get-configmap-cattle-controllers.yaml
Error from server (NotFound): configmaps "cattle-controllers" not found
command terminated with exit code 1
```
These are the configmaps I have in my cattle-system namespace:
```
$ kubectl get configmap -n cattle-system
NAME                           DATA   AGE
admincreated                   0      4h44m
rancher-partner-charts-nlqtw   1      4h44m
rancher-charts-dqswd           1      4h44m
```
The cattle-controllers configmap seems to be in my kube-system namespace:
```
$ kubectl get configmap -A | grep cattle-controllers
kube-system           cattle-controllers                            0      4h46m
```
Everything else looked fine.

mattmattox commented:
@dnoland1 I fixed the namespace for the configmap.

dkeightley commented:
👍  Could we add a README similar to the logs-collector before merging?

mattmattox commented:
@dkeightley I added a readme. I basically copied the one of logs-collector and changed the details.
I did add a section at the bottom called `Important disclaimer` that talks about trace loggings and how it can capture usernames, passwords, encryption keys, etc.

dkeightley commented:
Hey @mattmattox I added some changes to collect also the `cattle-system` namespace in the cluster-info dump (it defaults only to `kube-system`) and cluster objects as well. The output-directory style is a bit easier to navigate.

The cluster-info dump collects pod logs as well, so moved this to after the debug logging and removed the capture.

You want to review and modify if needed?

mattmattox commented:
@dnoland1 I had the wrong namespace for the configmap. It works now

mattmattox commented:
@dkeightley Looks awesome, thanks
